source/extensions/path/pattern_template_lib/pattern_template_internal.h

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <optional>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+#include "absl/types/variant.h"
+#include "re2/re2.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace PatternTemplate {
+
+namespace Internal {
+
+using Literal = absl::string_view;
+
+/**
+ * Determines what operations to use on the input pattern segment
+ */
+enum class Operator { KPathGlob, KTextGlob };
+
+/**
+ * Represents a segment of the rewritten URL, including any path segments,
+ * slash and prefix.
+ */
+struct RewriteSegment {
+
+  // A string that will be concatenated for rewrite.
+  absl::string_view literal_;
+
+  // Represents an index into the RE2 capture which value should be used
+  // to construct the rewritten URL. Note that the index should be greater
+  // than 0 as 0 index into the whole match RE2 pattern.
+  int capture_index_;
+};
```

Comment:
This struct seems to represent a captured segment, hence the `capture_index`. Is it really Rewrite specific? If not, perhaps we could call this CapturedSegment.